experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts

```diff
@@ -131,120 +139,131 @@ export const optionalChangeRebaser: FieldChangeRebaser<OptionalChangeset> = {
 		crossFieldManager: CrossFieldManager,
 		revisionMetadata: RevisionMetadataSource,
 	): OptionalChangeset => {
-		const perChildChanges = new ChildChangeMap<TaggedChange<NodeChangeset>[]>();
-		const addChildChange = (id: ChangeId, ...changeList: TaggedChange<NodeChangeset>[]) => {
-			const existingChanges = perChildChanges.get(id);
-			if (existingChanges !== undefined) {
-				existingChanges.push(...changeList);
-			} else {
-				perChildChanges.set(id, [...changeList]);
+		const getBidirectionalMaps = (
+			moves: OptionalChangeset["moves"],
+		): {
+			srcToDst: ChildChangeMap<[dst: RegisterId, target: "nodeTargeting" | "cellTargeting"]>;
+			dstToSrc: ChildChangeMap<RegisterId>;
+		} => {
+			const srcToDst = new ChildChangeMap<
+				[dst: RegisterId, target: "nodeTargeting" | "cellTargeting"]
+			>();
+			const dstToSrc = new ChildChangeMap<RegisterId>();
+			for (const [src, dst, target] of moves) {
+				srcToDst.set(src, [dst, target]);
+				dstToSrc.set(dst, src);
 			}
+			return { srcToDst, dstToSrc };
 		};
 
-		let fieldChange: Mutable<OptionalFieldChange> | undefined;
-		let currentChildNodeChanges: TaggedChange<NodeChangeset>[] = [];
-		let index = 0;
+		let latestReservedDetachId: RegisterId | undefined;
+		let isInputContextEmpty: boolean | undefined;
+
+		const builds: OptionalChangeset["build"] = [];
+		const childChangesByOriginalId = new ChildChangeMap<TaggedChange<NodeChangeset>[]>();
+		// TODO: It might be possible to compose moves in place rather than repeatedly copy.
+		// Additionally, working out a 'register allocation' strategy which enables frequent cancellation of noop moves
+		// for sandwich rebases would help with cloning if in-place proves too difficult
+		const current = getBidirectionalMaps([]);
 		for (const { change, revision } of changes) {
-			const fieldChangeInfo = revisionMetadata.tryGetInfo(
-				revision ?? change.fieldChange?.revision,
-			);
-			const { childChanges } = change;
-			if (childChanges !== undefined) {
-				for (const [childId, childChange] of childChanges) {
-					const taggedChildChange = tagChange(childChange, revision);
-					if (childId === "self") {
-						// childChange refers to the node that existed at the start of `change`,
-						// Thus in the composition, it should be referred to by whatever deletes that node in the future, which is what
-						// currentChildNodeChanges tracks
-						currentChildNodeChanges.push(taggedChildChange);
-					} else {
-						addChildChange(childId, taggedChildChange);
-					}
-				}
+			if (
+				isInputContextEmpty === undefined &&
+				(change.moves.length > 0 || change.reservedDetachId !== undefined)
+			) {
+				// change includes sets to the field, thus the emptiness of it previously can be inferred.
+				// Note that this block isn't entered for child-only changes.
+				isInputContextEmpty = change.reservedDetachId !== undefined;
 			}
```

Comment:
Consider refactoring into
```suggestion
			isInputContextEmpty ||= tryInferInputContext(change);
```
where `tryInferInputContext` would return `undefined` when `change.moves.length > 0 || change.reservedDetachId !== undefined` is false, and `change.reservedDetachId !== undefined` otherwise.
The main benefit I see for it is that making `tryInferInputContext` a separate function with a clear contract will makes it easier to assess its correctness (now and in the future).

Correctness-wise, it seems like a leap to infer that the presence of move entries implies that at least one of those moves affect the "self" register. There's at least nothing in the `OptionalChangeset` data model that guarantees this is the case. So some comment explainling the reasoning for this leap (or an actual run-time check) would be good.

Reply:
Thanks for the suggestion and prompting me to look into this a bit more--it bugged me too. Turns out that not making this logical leap was only blocked on generating `reservedDetachId`s for inverses (e.g. inverting `OptionalField.clear(wasEmpty: false, { localId: 0 })`), which is something we should be doing anyway.

Next update, the impl here should follow much more naturally from documented invariants on optional changesets.